packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -0,0 +1,95 @@
+use sqlx::{PgPool};
+use error_stack::{IntoReport, Result};
+use crate::datastore::{Datastore, DatastoreError};
+use crate::types::{GraphElementIdentifier, GraphElementType};
+
+/// A Postgres-backed Datastore
+pub struct PostgresDatabase {
+    pub pool: PgPool,
+}
+
+impl PostgresDatabase {
+    /// Creates a new `PostgresDatabase` object
+    ///
+    /// # Errors
+    /// If creating a [`PgPool`] connection returns an error
+    pub async fn new(connect_url: &str) -> Result<Self, sqlx::Error> {
+        Ok(Self {
+            pool: PgPool::connect(connect_url).await.report()?
+        })
+    }
+}
+
+#[allow(clippy::match_same_arms, reason="scaffolding while WIP")]
+impl Datastore for PostgresDatabase {
+    fn create(element: GraphElementType) -> Result<(), DatastoreError> {
+        match element {
+            GraphElementType::DataType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::PropertyType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::EntityType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::LinkType(_) => {
+                unimplemented!();
+            }
+        }
+    }
+
+    fn read_one(_id: GraphElementIdentifier) -> Result<GraphElementType, DatastoreError> {
+        unimplemented!()
+    }
+
+    fn read_many(_ids: &[GraphElementIdentifier]) -> Result<Vec<GraphElementType>, DatastoreError> {
+        unimplemented!()
+    }
+
+    fn update(_id: GraphElementIdentifier, element: GraphElementType) -> Result<(), DatastoreError> {
+        match element {
+            GraphElementType::DataType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::PropertyType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::EntityType(_) => {
+                unimplemented!();
+            }
+            GraphElementType::LinkType(_) => {
+                unimplemented!();
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const CONNECTION_STRING: &str = "postgres://postgres:postgres@localhost/postgres";
+
+    // TODO - long term we likely want to gate these behind config or something, probably do not
+    //  want to add a dependency on the external service for *unit* tests
+
+    #[tokio::test]
+    async fn can_connect() {
+        PostgresDatabase::new(CONNECTION_STRING)
+            .await
+            .expect("Couldn't connect to the Postgres DB");
+    }
+
+    #[tokio::test]
+    async fn get_entity_types() {
+        let db = PostgresDatabase::new(CONNECTION_STRING)
+            .await
+            .expect("Couldn't connect to the Postgres DB");
+
+        let entity_types = sqlx::query!("SELECT * from entity_types")
```

Comment:
This `query!` macro does some black-magic at compile-time where it actually connects to the database so that it can generate the structure of `entity_types`. This is pretty cool and I think we can leverage it for some more early-development but seems like a very bad idea in the long run, so I'm not against getting rid of it now. I just wanted to print out the result of a `select` to check/prove the simple DB method works.

Reply:
It's indeed a really cool feature. It's just annoying that a DB has to be running when you're compiling (unless we use `sqlx prepare` and cache the DB tables).
It's really helpful for developing, so perhaps we should encourage people to check with the macro, and then turn the call into the non-macro variant if it type checks. On the other hand, we could get some really cool CI checks if we boot up the DB and allow the checks